pennylane/transforms/hamiltonian_expand.py

```diff
@@ -129,20 +129,26 @@ def hamiltonian_expand(tape, group=True):
             "Passed tape must end in `qml.expval(H)`, where H is of type `qml.Hamiltonian`"
         )
 
+    # note: for backward passes of some frameworks
+    # it is crucial to use the hamiltonian.data attribute,
+    # and not hamiltonian.coeffs when recombining the results
+
     if group or hamiltonian.grouping_indices is not None:
 
         if hamiltonian.grouping_indices is None:
+            # explicitly selected grouping, but indices not yet computed
             hamiltonian.compute_grouping()
 
-        # use groups of observables if available or explicitly requested
-        coeffs = [
-            qml.math.squeeze(qml.math.take(hamiltonian.coeffs, indices, axis=0))
+        coeff_groupings = [
+            qml.math.stack([hamiltonian.data[i] for i in indices])
             for indices in hamiltonian.grouping_indices
         ]
```

Comment:
For example, for two groups of the form `[[PauliX(0), PauliX(1)], [PauliZ(0)]]` and tensorflow, `coeff_groupings` looks like this:
`[<tf.Tensor: shape=(2,), dtype=float64, numpy=array([-0.05,  1.2 ])>, <tf.Tensor: shape=(1,), dtype=float64, numpy=array([0.17])>]`
and the `res_groupings` below used in post processing like this:
`[<tf.Tensor: shape=(2,), dtype=float64, numpy=array([-0.12777055,  1.        ])>, <tf.Tensor: shape=(1,), dtype=float64, numpy=array([0.0166009])>]`